New: button text +/ icon support, items removed and a11y enhancements (fixes #13)

kirsty-hames opened the pull request:
This PR addresses the following tasks raised in issue #13:


**1. Switch to notify only with text + graphic in schema**
- native popup removed and notify popup made default
- a11y titles (`title` and `altTitle`) setup as per notify
Fixes https://github.com/cgkineo/adapt-hint/issues/14 (now consistent with notify a11y titles, item titles void as these have been removed)
- _notify graphic support still outstanding - raised as a separate issue to gain further input https://github.com/cgkineo/adapt-hint/issues/16_ 

**2. Position under component header and before widget**
- append the Hint button between the component display text and the widget to suite the typical use case
- remove the absolute position so the button display matches the DOM order (WCAG C27: Making the DOM order match the visual order)
Fixes https://github.com/cgkineo/adapt-hint/issues/7 
Fixes https://github.com/cgkineo/adapt-hint/issues/6 (button is no longer a direct child of component so issue is void)


**3. Allow button text +/ icon**
- make icon class configurable
- make icon alignment configurable
- add component level text and ariaLabel (as well as existing global aria)
- remove button border-radius and padding styles as these are inherited from `.btn-icon` and `.btn-text` selectors
- new properties and globals added to example


**4. remove items (only single title and body are used)**
- on reviewing use case, only single title and body are used per hint (multiple items aren't being used)
- title inherits notify title style
- replace `.hint-popup__item` with `.hint-popup__content` containers for consistency with other notify popups (hotgraphic and tutor)

Fixes https://github.com/cgkineo/adapt-hint/issues/13

# Testing PR
Apply the following config to any component. Typical use cases are:
- 'hints' for question components (providing additional content to aid a user in answering a question)
- 'more info' for presentation components (to stash away text heavy content, or, content of a particular type/theme)

```
    "_hint": {
        "title": "Adapt Hint",
        "altTitle": "Alt Hint title text",
        "body": "An extension that adds a small, clickable icon to a component that displays additional information.",
        "_button": {
            "_iconClass": "icon-question",
            "_alignIconRight": true,
            "text": "Hint",
            "ariaLabel": ""
        }
    }
```
And the globals in course.json:
```
    "_globals": {
        "_hint": {
            "openButtonText": "Open hint"
        }
    }
```

# Notes
- I have a11y tested this with VoiceOver on Mac
- needs AAT testing
- README demo.gif needs replacing to reflect new functionality/display.

kirsty-hames committed:
remove native popup and use notify popup only

- _isNotifyPopup config removed and notify made default
- closeButtonText removed as this was only applicable to the native popup

kirsty-hames committed:
remove items

- only single title and body are used per hint
- title inherits notify title style
- replace hint-popup__item with hint-popup__content containers for consistency with other notify popups (hotgraphic and tutor)

kirsty-hames committed:
add popup class

- remove notify popup defaults

kirsty-hames committed:
replace ariaLabelNotify with altTitle

- use altTitle for consistency with other notify popups

kirsty-hames committed:
replace margin selector with css style

kirsty-hames committed:
remove redundant js selector

kirsty-hames committed:
position button between DT and widget

- for a11y, remove absolute position so the visual order matches the DOM order

kirsty-hames committed:
remove native popup styling

...native popup removed in previous commit

kirsty-hames committed:
text and/or icon support for button

- make icon class configurable
- make icon alignment configurable
- add component level text and ariaLabel (as well as existing global aria)
- remove button border-radius and padding styles as these are inherited from btn-icon and btn-text selectors
- new properties and globals added to example
- schema needs updating!

-

kirsty-hames committed:
popup styling

- add notify link styling

kirsty-hames committed:
add margin between button and widget

kirsty-hames committed:
schema updates

kirsty-hames committed:
bower updates

kirsty-hames committed:
schema $anchor fixes

oliverfoster commented on js/adapt-hint.js:
```diff
@@ -22,27 +22,21 @@ define([
       const template = Handlebars.templates.hint;
 
-      this.$el.html(template(data)).appendTo($('.' + this.model.get('_id')));
+      this.$el.html(template(data)).appendTo($('.' + this.model.get('_id')).find('.component__header-inner'));
```
Do we need a fallback if the header isn't available?

kirsty-hames replied:
Fallback appends hint to the component widget instead (commit [1b30b7b](https://github.com/cgkineo/adapt-hint/commit/1b30b7b41abf88067f524764a9e223362ecfddac)).

kirsty-hames commented on properties.schema:
```diff
@@ -6,24 +6,9 @@
   "globals": {
     "openButtonText": {
```
Should `openButtonText` be nested within `_extensions`?

oliverfoster replied:
Yes should, the schema applies from the root of the course json.

kirsty-hames committed:
example _globals updated

kirsty-hames committed:
fallback if the component header isn't available

...append to the component widget instead

kirsty-hames committed:
Update README.md

...gif and versioning outstanding

kirsty-hames committed:
remove version support from README

kirsty-hames commented:
Priority for this PR is testing in the AAT (legacy) as this is required for a project. I haven't been able to upload this to an AAT yet and getting the following error on attempting upload: 

_Plugin upload failed
Error: Failed to parse schema for package adapt-hint_

Is this just a case of a missing _package.json_?

joe-allen-89 committed:
Fix: package files added, typo in schema amended

joe-allen-89 commented:
> Priority for this PR is testing in the AAT (legacy) as this is required for a project. I haven't been able to upload this to an AAT yet and getting the following error on attempting upload:
> 
> _Plugin upload failed Error: Failed to parse schema for package adapt-hint_
> 
> Is this just a case of a missing _package.json_?

@kirsty-hames There was a couple of rogue commas in the schema file causing this issue, that's been amended. I've also added the package files whilst I was at it.

kirsty-hames committed:
Update bower description

kirsty-hames committed:
Add _isEnabled

...to prevent the Hint button being rendered for every component (testing in AAT).

Button ARIA label description added.

Bower description updated.

kirsty-hames committed:
bower and package json description updated
